targets/TARGET_Silicon_Labs/TARGET_EFM32/trng/sl_trng.c

```diff
@@ -0,0 +1,274 @@
+    device->CONTROL = ctrl;
+}
+
+static void sl_trng_clear_fifo( TRNG_TypeDef *device )
+{
+  volatile uint32_t val32;
+
+  /* Empty FIFO */
+  while ( device->FIFOLEVEL )
+  {
+      val32 = device->FIFO;
+      (void)val32;
+  }
+}
+
+int sl_trng_set_key( TRNG_TypeDef *device, const unsigned char *key )
+{
+    uint32_t *_key = (uint32_t*) key;
+
+    sl_trng_clear_fifo(device);
+
+    /* Program key in KEY registers of the TRNG. */
+    device->KEY0 = *_key++;
+    device->KEY1 = *_key++;
+    device->KEY2 = *_key++;
+    device->KEY3 = *_key++;
+
+    return 0;
+}
+
+static int sl_trng_check_status( TRNG_TypeDef *device )
+{
+    uint32_t status = device->STATUS;
+
+    if ( (status & (TRNG_STATUS_PREIF
+                    | TRNG_STATUS_REPCOUNTIF
+                    | TRNG_STATUS_APT64IF
+                    | TRNG_STATUS_APT4096IF
+                    | TRNG_STATUS_ALMIF)) == 0 )
+    {
+        /* No errors */
+        return 0;
+    }
+
+    if ( status & TRNG_STATUS_PREIF )
+    {
+        /* On a preliminary noise alarm we clear the FIFO and clear
+         * the alarm. The preliminary noise alarm is not critical. */
+        status &= ~TRNG_STATUS_PREIF;
+        device->STATUS = status;
+        sl_trng_clear_fifo(device);
+        return SL_TRNG_ERR_PRELIMINARY_NOISE_ALARM;
+    }
+    else
+    {
+        /* Clear alarm conditions by doing a TRNG soft reset. */
+        sl_trng_soft_reset( device );
+        if ( status & TRNG_STATUS_REPCOUNTIF )
+        {
+            return SL_TRNG_ERR_REPETITION_COUNT_TEST_FAILED;
+        }
+        if ( status & TRNG_STATUS_APT64IF )
+        {
+            return SL_TRNG_ERR_ADAPTIVE_PROPORTION_TEST_64_FAILED;
+        }
+        if ( status & TRNG_STATUS_APT4096IF )
+        {
+            return SL_TRNG_ERR_ADAPTIVE_PROPORTION_TEST_4096_FAILED;
+        }
+        if ( status & TRNG_STATUS_ALMIF )
+        {
+            return SL_TRNG_ERR_NOISE_ALARM;
+        }
+    }
+
+    return 0;
+}
+
+static void sl_trng_read_chunk( TRNG_TypeDef *device,
+                                unsigned char *output,
+                                size_t len )
+{
+    uint32_t * out32 = (uint32_t *) output;
+    uint32_t tmp;
+
+    /* Read known good available data. */
+    while ( len >= 4)
+    {
+        *out32++ = device->FIFO;
+        len -= 4;
+    }
+
+    /* Handle the case where len is not a multiple of 4. */
+    if ( len < 4 )
+    {
+        tmp = device->FIFO;
+        memcpy((uint8_t *)out32, (const uint8_t *) &tmp, len);
+    }
+}
+
+int sl_trng_poll( TRNG_TypeDef *device,
+                  unsigned char *output,
+                  size_t len,
+                  size_t *olen )
+{
+    size_t output_len = 0;
+    size_t chunk_len = 0;
+    size_t available;
+    int ret = 0;
+
+    while (output_len < len)
+    {
+        available = device->FIFOLEVEL * 4;
+        if (available == 0)
+        {
+            break;
+        }
+
+#if !defined(SL_TRNG_IGNORE_ALL_ALARMS)
```

Comment:
Could somebody please help me understand why we need these alarm suppression options? (I mean `SL_TRNG_IGNORE_ALL_ALARMS` and `SL_TRNG_IGNORE_NOISE_ALARMS`)